File: wc.go
package main

import "fmt"
import "flag"
import "os"
import "io"
import "log"
import "regexp"
import "strings"
import "unicode/utf8"

func main() {
	var c = flag.Bool("c", false, "count bytes")
	var l = flag.Bool("l", false, "count lines")
	var m = flag.Bool("m", false, "count chars")
	var w = flag.Bool("w", false, "count words")
	flag.Parse()

	if *m {
		*c = false
	}

	if !*c && !*l && !*m && !*w {
		*c = true
		*l = true
		*w = true
	}

	var contents []byte
	var err error
	filename := ""
	if flag.NArg() > 0 {
		filename = flag.Arg(0)
		contents, err = os.ReadFile(filename)
	} else {
		contents, err = io.ReadAll(os.Stdin)
	}

	if err != nil {
		log.Panic(err)
	}

	result := make([]int, 0, 3)
	text := ""

	if *l || *w || *m {
		text = string(contents)
	}

	if *l {
		trimmed := strings.TrimRight(text, "\n")
		result = append(result, len(regexp.MustCompile("\r?\n").Split(trimmed, -1)))
	}

	if *w {
		result = append(result, len(strings.Fields(text)))
	}

	if *c {
		result = append(result, len(contents))
	}

	if *m {
		result = append(result, utf8.RuneCountInString(text))
	}

	fmt.Println("\t", strings.Trim(strings.Join(strings.Fields(fmt.Sprint(result)), "\t"), "[]"), filename)
}

File: wc.hs
import Data.List
import qualified Data.Text as T
import qualified Data.Text.Encoding as T
import qualified Data.ByteString as BS
import System.Environment
import System.Exit
import System.Console.GetOpt
import System.IO

data Flag = Lines | Words | Bytes | Chars
  deriving (Eq, Ord, Enum, Show, Bounded)

flags :: [OptDescr Flag]
flags = [ Option ['c'] [] (NoArg Bytes) "count bytes"
        , Option ['m'] [] (NoArg Chars) "count chars"
        , Option ['w'] [] (NoArg Words) "count words"
        , Option ['l'] [] (NoArg Lines) "count lines"
        ]

gatherFlags :: [Flag] -> [Flag]
gatherFlags args
  | null args = [Lines, Words, Bytes]
  | Bytes `elem` args && Chars `elem` args = filter (\f -> f /= Bytes) args
  | otherwise = args

parse :: [String] -> IO ([Flag], [String])
parse argv = case getOpt Permute flags argv of
  (args, fs, []) -> do
    let files = if null fs then [] else fs
    return (gatherFlags args, files)
  (_,_,errs) -> do
    hPutStrLn stderr (concat errs ++ usageInfo header flags)
    exitWith (ExitFailure 1)
  where header = "Usage: wc [-clmw] file"

readInput :: [String] -> IO BS.ByteString
readInput [] = BS.getContents
readInput files = BS.readFile $ head files

countStuff :: Flag -> BS.ByteString -> T.Text -> Int
countStuff Chars _ = T.length
countStuff Lines _ = length . T.lines
countStuff Words _ = length . T.words
countStuff Bytes s = const $ BS.length s

needsDecode :: [Flag] -> Bool
needsDecode = any (/= Bytes)

-- lwcm
main :: IO ()
main = do
  (args, files) <- getArgs >>= parse
  contents <- readInput files
  let text = if needsDecode args
        then T.decodeUtf8 contents
        else T.empty
  let result = [ countStuff f contents text | f <- sort args]
  putStr $ "\t" ++ (intercalate "\t" $ map show result)
  if not $ null files
    then putStrLn $ " " ++ head files
    else putStrLn ""
